packages/keyring-controller/src/KeyringController.ts

```diff
@@ -1724,81 +1725,94 @@ export class KeyringController extends BaseController<
     encryptionKey?: string,
     encryptionSalt?: string,
   ): Promise<EthKeyring<Json>[]> {
-    const encryptedVault = this.state.vault;
-    if (!encryptedVault) {
-      throw new Error(KeyringControllerError.VaultError);
-    }
+    return this.#withVaultLock(async ({ releaseLock }) => {
+      const encryptedVault = this.state.vault;
+      if (!encryptedVault) {
+        throw new Error(KeyringControllerError.VaultError);
+      }
 
-    await this.#clearKeyrings();
+      await this.#clearKeyrings({ skipStateUpdate: true });
```

Comment:
If the function throws an error before committing to the state/vault, the controller will be left in an inconsistent state that could cause unwanted scenarios. We should refactor the unlock operation to be completely atomic

Reply:
Hmm. Ideally, all operations should be atomic: 
e.g. when adding an account to a keyring, any error thrown _after_ the account addition on the keyring and _before_ the state/vault update should not leave the new account on the Keyring. But this might be out of scope for this PR

Reply:
This could become more concerning when messenger events fired after an update operation come into play: afaik, if some listener to state changes throws an error, it will block the rest of the KeyringController function that published the event in the first place, most likely jumping the vault update step